ports/rp2/machine_uart.c

```diff
@@ -335,7 +335,13 @@ STATIC void machine_uart_init_helper(machine_uart_obj_t *self, size_t n_args, co
         }
 
         uart_init(self->uart, self->baudrate);
+        // According to the comment in the implementation of uart_init()
+        // the UART must not be enabled for changes to the LCR register.
+        // Thus, it is disabled and enabled again after the format setting.
+        uart_get_hw(self->uart)->cr &= ~(UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS);
         uart_set_format(self->uart, self->bits, self->stop, self->parity);
+        uart_get_hw(self->uart)->cr = UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS;
```

Comment:
Should this be `|=`?

Reply:
Can be. But at that flow of the code after the call of uart_init() these are anyhow the only bits being set, so there is no difference. But I can change it.

Reply:
It should be `|=` to match the `&= ~`.

Otherwise it could just be ` = 0` instead of the `&= ~`.

Reply:
But retesting the change I ran into another strange behavior. After reset, instantiating the port once using parity fails. Pariyt is still wrong. Instantiating UART twice works. So this "simple" change requires more testing.

Reply:
Might be worth pinging -@- kilograham to ask for advice, when we have a specific question to ask.

Reply:
Using the first choice with a delay after settinf parity etc. works. Hmmm..

Reply:
Note it seems `uart_set_fifo_enabled` also changes `LCR`, if that comment about Not changing `LCR` when UART is enabled is true, then you should move the enable after `uart_set_fifo_enabled`, i.e.:

```c
uart_get_hw(self->uart)->cr &= ~(UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS);
uart_set_format(self->uart, self->bits, self->stop, self->parity);
uart_set_fifo_enabled(self->uart, true);
uart_get_hw(self->uart)->cr |= UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS;
```

Also it might be enough to just toggle this bit `UART_UARTCR_UARTEN_BITS`

Reply:
It seems that just dropping `uart_set_fifo_enabled()` after `uart_set_format()` solves the problem. FIFO is already enabled in uart_init(). Using a debugger and testing I could verify, that then after the call to `uart_set_format()` all bits are properly set.

Reply:
> Note it seems uart_set_fifo_enabled also changes LCR, if that comment about Not changing LCR when UART is enabled is true,

If you look at the code snippet below, which is from uart_init(), you see that even there lcr_h is modified AFTER the UART was enabled.

 ```
    // Any LCR writes need to take place before enabling the UART
    uint baud = uart_set_baudrate(uart, baudrate);
    uart_set_format(uart, 8, 1, UART_PARITY_NONE);

    // Enable the UART, both TX and RX
    uart_get_hw(uart)->cr = UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS;
    // Enable FIFOs
    hw_set_bits(&uart_get_hw(uart)->lcr_h, UART_UARTLCR_H_FEN_BITS);
    // Always enable DREQ signals -- no harm in this if DMA is not listening
```

Reply:
Maybe you just need `DSB` after `LCR_H` writes ?

Reply:
That seems to work as well, even with the call to `uart_set_fifo_enabled()`. But the call to `uart_set_fifo_enabled()` is not required, since FIFO is already enabled. And then the `__DSB()` is not needed. It could be added, but if needed it should be better in the `uart_set_format()` code.

Reply:
> but if needed it should be better in the uart_set_format() code

I agree.